src/rulesets/Base/mapreduce.jl

```diff
@@ -56,6 +77,17 @@ function rrule(
     return y, sum_abs2_pullback
 end
 
+# Fix dispatch for this pidgeon-hole optimization,
+# Rules with RuleConfig dispatch with priority over without (regardless of other args).
+# and if we don't specify what do do for one that HasReverseMode then it is ambigious
+for Config in (RuleConfig, RuleConfig{>:HasReverseMode})
+    @eval function rrule(
+        ::$Config, ::typeof(sum), ::typeof(abs2), x::AbstractArray{T}; dims=:,
+    ) where {T<:Union{Real,Complex}}
+        return rrule(sum, abs2, x; dims=dims)
+    end
+end
```

Comment:
I don't love this.
But I think it is as good as it gets.
We could require every rule is written to have to have a `RuleConfig`,
but we would still get the ambiguity, and so would still need the same code (just with one less thing in the top loop.

If we are happy with this I will make a follow up PR to the ChainRulesCore docs, warning about this.

Reply:
Eww. But I can't think of anything nicer either. Can we add tests for this as well?

Reply:
Why would `
rrule(config::RuleConfig{&gt;:HasReverseMode}, ::typeof(sum), f::typeof(abs2), xs::AbstractArray)` not be more specific?

Reply:
It is, which is why we have to generate that.

The problem is with:
```
rrule(config::RuleConfig, ::typeof(sum), f::typeof(abs2), xs::AbstractArray)
```
which is what we could make the required format.
That one is ambiguous with
```
rrule(config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray)
```